mozilla_django_oidc/utils.py

```diff
@@ -51,3 +57,54 @@ def is_authenticated(user):
     msg = '`is_authenticated()` is going to be removed in mozilla-django-oidc v 2.x'
     warnings.warn(msg, DeprecationWarning)
     return user.is_authenticated
+
+
+def add_state_and_nonce_to_session(request, state, params):
+    """
+    Stores the `state` and `nonce` parameters in a session dictionary including the time when it
+    was added. The dictionary can contain multiple state/nonce combinations to allow parallel
+    logins with multiple browser sessions.
+    To keep the session space to a reasonable size, the dictionary is kept at 50 state/nonce
+    combinations maximum.
+    """
+    nonce = None
+    if import_from_settings('OIDC_USE_NONCE', True):
+        nonce = get_random_string(import_from_settings('OIDC_NONCE_SIZE', 32))
+        params.update({
+            'nonce': nonce
+        })
+
+    # Store Nonce with the State parameter in the session "oidc_states" dictionary.
+    # The dictionary can store multiple State/Nonce combinations to allow parallel
+    # authentication flows which would otherwise overwrite State/Nonce values!
+    # The "oidc_states" dictionary uses the state as key and as value a dictionary with "nonce"
+    # and "added_on". "added_on" contains the time when the state was added to the session.
+    # With this value, the oldest element can be found and deleted from the session.
+    if 'oidc_states' not in request.session or \
+            not isinstance(request.session['oidc_states'], dict):
+        request.session['oidc_states'] = {}
+
+    # Make sure that the State/Nonce dictionary in the session does not get too big.
+    # If the number of State/Nonce combinations reaches a certain threshold, remove the oldest
+    # state by finding out
+    # which element has the oldest "add_on" time.
+    limit = import_from_settings('OIDC_MAX_OIDC_STATES', 50)
+    limit = import_from_settings('OIDC_MAX_STATES', 50)
+    if len(request.session['oidc_states']) >= limit:
+        LOGGER.warning(
```

Comment:
Nit: I think this should be an `.INFO` level log, but maybe we don't expect it in production, only if `OIDC_MAX_STATES` is too low?